tree/dataframe/inc/ROOT/RDF/ActionHelpers.hxx

```diff
@@ -377,142 +377,116 @@ public:
 
    void InitTask(TTreeReader *, unsigned int) {}
 
-   void Exec(unsigned int slot, double x0) // 1D histos
+   // all scalar arguments
+   template <typename... ValTypes,
+             typename std::enable_if<Conjunction<std::is_convertible<ValTypes, double>...>::value, int>::type = 0>
+   void Exec(unsigned int slot, ValTypes... x)
    {
-      fObjects[slot]->Fill(x0);
+      fObjects[slot]->Fill(x...);
    }
 
-   void Exec(unsigned int slot, double x0, double x1) // 1D weighted and 2D histos
-   {
-      fObjects[slot]->Fill(x0, x1);
-   }
+   // helper functions which provide one implementation for scalar types and another for containers
+   // TODO these could probably all be replaced by inlined lambdas and/or constexpr if statements
+   // in c++17 or later
 
-   void Exec(unsigned int slot, double x0, double x1, double x2) // 2D weighted and 3D histos
+   // return unchanged value for scalar
+   template <typename T, typename std::enable_if<std::is_convertible<T, double>::value, int>::type = 0>
+   const T &MakeBegin(const T &val)
    {
-      fObjects[slot]->Fill(x0, x1, x2);
+      return val;
    }
 
-   void Exec(unsigned int slot, double x0, double x1, double x2, double x3) // 3D weighted histos
+   // return iterator to beginning of container
+   template <typename T, typename std::enable_if<!std::is_convertible<T, double>::value, int>::type = 0>
+   typename T::const_iterator MakeBegin(const T &val)
    {
-      fObjects[slot]->Fill(x0, x1, x2, x3);
+      return std::begin(val);
    }
 
-   template <typename X0, std::enable_if_t<IsDataContainer<X0>::value || std::is_same<X0, std::string>::value, int> = 0>
-   void Exec(unsigned int slot, const X0 &x0s)
+   // return unchanged value for scalar
+   template <typename T, typename std::enable_if<std::is_convertible<T, double>::value, int>::type = 0>
+   const T &Increment(const T &val)
    {
-      auto thisSlotH = fObjects[slot];
-      for (auto x0 = x0s.begin(); x0 != x0s.end(); x0++) {
-         thisSlotH->Fill(*x0); // TODO: Can be optimised in case T == vector<double>
-      }
+      return val;
    }
 
-   // ROOT-10092: Filling with a scalar as first column and a collection as second is not supported
-   template <typename X0, typename X1,
-             std::enable_if_t<IsDataContainer<X1>::value && !IsDataContainer<X0>::value, int> = 0>
-   void Exec(unsigned int, const X0 &, const X1 &)
+   // return incremented reference for iterators
+   template <typename T, typename std::enable_if<!std::is_convertible<T, double>::value, int>::type = 0>
+   T &Increment(T &val)
    {
-      throw std::runtime_error(
-        "Cannot fill object if the type of the first column is a scalar and the one of the second a container.");
```

Comment:
We used to disallow the combination of scalar value and array weights, is it allowed with these changes? I need to go back to the discussion on Jira and see what we wanted to do there

Reply:
This is allowed with these changes.  There is even an explicit use case for this, where a vector of weights represents a systematic variation, albeit the typical use case is something like a TH2D with one scalar value, one vector value for the systematic index, and a vector value for the systematic weights.

Reply:
Ok I'll bring this up with @hageboeck and others that were involved in the discussion at https://sft.its.cern.ch/jira/browse/ROOT-10486 -- it's not a blocker for the PR but we want to make up our minds before the release.